app/src/org/commcare/android/database/app/models/FormDefRecord.java

```diff
@@ -0,0 +1,183 @@
+    }
+
+    public int save(SqlStorage<FormDefRecord> formDefRecordStorage) {
+        // if we don't have a path to the file, the rest are irrelevant.
+        // it should fail anyway because you can't have a null file path.
+        if (StringUtils.isEmpty(mFormFilePath)) {
+            throw new SQLException("Can't save a form def with an empty form file path");
+        }
+        // Make sure that the necessary fields are all set
+
+        File form = new File(mFormFilePath);
+        if (StringUtils.isEmpty(mDisplayName)) {
+            mDisplayName = form.getName();
+        }
+
+        if (StringUtils.isEmpty(mFormMediaPath)) {
+            mFormMediaPath = getMediaPath(mFormFilePath);
+        }
+
+        formDefRecordStorage.write(this);
+
+        if (recordId == -1) {
+            throw new SQLException("Failed to save the FormDefRecord " + toString());
+        }
+        return recordId;
+    }
+
+    private String getMediaPath(String formFilePath) {
+        String pathNoExtension = formFilePath.substring(0, formFilePath.lastIndexOf("."));
+        return pathNoExtension + "-media";
+    }
+
+    public static void updateFilePath(SqlStorage<FormDefRecord> formDefRecordStorage, int recordId, String formFilePath) {
+        FormDefRecord existingRecord = formDefRecordStorage.read(recordId);
+        existingRecord.updateFilePath(formDefRecordStorage, formFilePath);
+    }
+
+
+    public void updateFilePath(SqlStorage<FormDefRecord> formDefRecordStorage, String newFilePath) {
+        File newFormFile = new File(newFilePath);
+
+        try {
+            if (new File(mFormFilePath).getCanonicalPath().equals(newFormFile.getCanonicalPath())) {
+                // Files are the same, so we may have just copied over something we had already
+            } else {
+                // New file name. This probably won't ever happen, though.
+                FileUtil.deleteFileOrDir(mFormFilePath);
+            }
+        } catch (IOException ioe) {
+            //we only get here if we couldn't canonicalize, in which case we can't risk deleting the old file
+            //so don't do anything.
+        }
+
+        // Set new values now
+        mFormMediaPath = getMediaPath(newFilePath);
+        formDefRecordStorage.write(this);
+    }
+
+    public static FormDefRecord getFormDef(SqlStorage<FormDefRecord> formDefRecordStorage, int formId) {
+        return formDefRecordStorage.read(formId);
+    }
+
+    public String getFilePath() {
+        return mFormFilePath;
+    }
+
+    public String getMediaPath() {
+        return mFormMediaPath;
+    }
+
+    public String getJrFormId() {
+        return mJrFormId;
+    }
+
+    public String getDisplayname() {
+        return mDisplayName;
+    }
+
+    public Integer getModelVersion() {
+        return mModelVersion;
+    }
+
+    public Integer getUiVersion() {
+        return mUiVersion;
+    }
+
+    public String getBase64RsaPublicKey() {
```

Comment:
Are these for the migration only? If so it would be useful to visible separate them out, or to use a subclass exposing these to allow them to be read only for the migration or some other technique to prevent confusion

Reply:
It's getting used in Encryption currently. Migration doesn't use any of these getters.

Reply:
I think this goes along with my comment above. If this is part of the EncryptedFormInformation code, i don't think we use this